### PR TITLE
Fix This instantiation

### PR DESCRIPTION
**Pull request by varunagrawal**
We can now use `This` with non-templated classes as well.

I am not sure if that is what @ProfFan was hinting at with the last task item in https://github.com/borglab/gtsam/issues/407 but it works for both templated an non-templated classes. :slightly_smiling_face:

### CODE REVIEW

**Comment by varunagrawal** [comment]
Still need to add a test.

**Comment by varunagrawal** [comment]
Let's land #55 before this since then I can update the tests accordingly.

File: gtwrap/template_instantiator.py [hunk 1]
```diff
@@ -272,28 +273,27 @@ def __init__(self, original, instantiations=(), new_name=''):
...
         self.parent = original.parent
 
-        if not original.template:
-            self.name = original.name
-            self.ctors = list(original.ctors)
-            self.static_methods = list(original.static_methods)
-            class_instantiated_methods = list(original.methods)
-            self.properties = list(original.properties)
-        else:
+        if original.template:
```
Comment (dellaert) [inline gtwrap/template_instantiator.py:276]: Now that you know what this code does, and logic below, add a comment here to that effect.

File: gtwrap/template_instantiator.py [hunk 2]
```diff
@@ -272,28 +273,27 @@ def __init__(self, original, instantiations=(), new_name=''):
...
-                original.name, instantiations) if not new_name else new_name
-            self.ctors = self.instantiate_ctors()
-            self.static_methods = self.instantiate_static_methods()
-            class_instantiated_methods = \
-                self.instantiate_class_templates_in_methods()
-            self.properties = self.instantiate_properties()
+        self.name = instantiate_name(
+            original.name, instantiations) if not new_name else new_name
+
+        # Check for typenames if templated.
```
Comment (dellaert) [inline gtwrap/template_instantiator.py:284]: nice!

File: gtwrap/template_instantiator.py [hunk 3]
```diff
@@ -328,13 +328,19 @@ def __repr__(self):
...
             )
 
-    def instantiate_ctors(self):
-        """Instantiate the class constructors."""
+    def instantiate_ctors(self, typenames):
+        """
+        Instantiate the class constructors.
+        
+        Args:
+            typenames: List of template types to instantiate.
```
Comment (dellaert) [inline gtwrap/template_instantiator.py:336]: @return?

File: gtwrap/template_instantiator.py [hunk 4]
```diff
@@ -345,13 +351,18 @@ def instantiate_ctors(self):
...
         return instantiated_ctors
 
-    def instantiate_static_methods(self):
-        """Instantiate static methods in the class."""
+    def instantiate_static_methods(self, typenames):
+        """
+        Instantiate static methods in the class.
+
+        Args:
+            typenames: List of template types to instantiate.
```
Comment (dellaert) [inline gtwrap/template_instantiator.py:359]: @return?

File: gtwrap/template_instantiator.py [hunk 5]
```diff
@@ -371,17 +382,20 @@ def instantiate_static_methods(self):
...
-    def instantiate_class_templates_in_methods(self):
+    def instantiate_class_templates_in_methods(self, typenames):
         """
         This function only instantiates class templates in the methods.
         Template methods are instantiated in InstantiatedMethod in the second
         round.
+
+        Args:
+            typenames: List of template types to instantiate.
         """
```
Comment (dellaert) [inline gtwrap/template_instantiator.py:393]: etc..

File: gtwrap/template_instantiator.py [hunk 6]
```diff
@@ -400,11 +414,16 @@ def instantiate_class_templates_in_methods(self):
...
 
-    def instantiate_properties(self):
-        """Instantiate the class properties."""
+    def instantiate_properties(self, typenames):
+        """
+        Instantiate the class properties.
+
+        Args:
+            typenames: List of template types to instantiate.
+        """
```
Comment (dellaert) [inline gtwrap/template_instantiator.py:423]: @return ?

**Review by dellaert — APPROVED** [review]
Maybe a bit more docs now that you're in this

**Comment by varunagrawal** [comment]
Honestly, there is so much more documentation left to be done. I'll add more.

**Comment by varunagrawal** [comment]
Added the test, will add docs after I get some physical activity.

**Comment by varunagrawal** [comment]
Added docs. Merging.

**Pull request merged**